storage/src/mmr/hasher.rs

```diff
@@ -1,82 +1,213 @@
-        self.update_with_hash(right_hash);
-        self.finalize_reset()
+        self.update_with_digest(left);
+        self.update_with_digest(right);
+        self.finalize()
     }
 
-    /// Computes the root hash for an MMR given its size and an iterator over the hashes of its
-    /// peaks. The iterator should yield the peak hashes in decreasing order of their height.
-    pub(crate) fn root_hash<'b>(
+    fn root_digest<'b>(
         &mut self,
         pos: u64,
-        peak_hashes: impl Iterator<Item = &'b H::Digest>,
+        peak_digests: impl Iterator<Item = &'b H::Digest>,
     ) -> H::Digest {
         self.update_with_pos(pos);
-        for hash in peak_hashes {
-            self.update_with_hash(hash);
+        for digest in peak_digests {
+            self.update_with_digest(digest);
         }
-        self.finalize_reset()
+        self.finalize()
     }
 
-    pub(crate) fn update_with_pos(&mut self, pos: u64) {
-        self.hasher.update(&pos.to_be_bytes());
+    fn digest(&mut self, data: &[u8]) -> H::Digest {
+        self.hasher.update(data);
+        self.finalize()
     }
-    pub(crate) fn update_with_hash(&mut self, hash: &H::Digest) {
-        self.hasher.update(hash.as_ref());
+}
+
+/// Hasher for computing leaf, node and root digests when the tree is being grafted onto another
+/// MMR.
+pub struct Grafting<'a, H: CHasher, S: Storage<H::Digest>> {
+    hasher: Basic<'a, H>,
+    height: u32,
+    base_mmr: &'a S,
+}
+
+impl<'a, H: CHasher, S: Storage<H::Digest>> Grafting<'a, H, S> {
+    pub fn new(hasher: &'a mut H, height: u32, base_mmr: &'a S) -> Self {
+        Self {
+            hasher: Basic::new(hasher),
+            height,
+            base_mmr,
+        }
     }
-    pub(crate) fn update_with_element(&mut self, element: &[u8]) {
-        self.hasher.update(element);
+
+    /// Access the underlying Basic (non-grafting) hasher.
+    pub fn basic(&mut self) -> &mut Basic<'a, H> {
+        &mut self.hasher
     }
-    pub(crate) fn finalize_reset(&mut self) -> H::Digest {
-        self.hasher.finalize()
+
+    /// Compute the position of the leaf in the base tree onto which we should graft the leaf at
+    /// position `pos` in the source tree.
+    ///
+    /// This position is computed by walking up the MMR exactly self.height steps from the leaf.
+    /// Since we don't know exactly where in the subtree this leaf falls, we map it to the
+    /// right-most leaf in the subtree and walk up from there. Runtime is O(log2(n)) in the given
+    /// position.
+    fn destination_pos(&self, pos: u64) -> u64 {
+        let peak_leaf_num = leaf_pos_to_num(pos).unwrap();
+        let chunk_size_bits = 1 << self.height;
+
+        // The rightmost-leaf in the corresponding segment of the peak tree:
+        let base_leaf_num = (peak_leaf_num * chunk_size_bits) + chunk_size_bits - 1;
+
+        // Walking up self.height levels from the rightmost leaf involves simply adding self.height
```

Comment:
nit: I feel this comment is better attributed to the following line (and should be moved or the `\n` between should be deleted)

Reply:
Got rid of the "let" entirely since it's not used for anything but the returned result.